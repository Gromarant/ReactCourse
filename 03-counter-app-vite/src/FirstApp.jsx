//lesson 40 - Components comunication

  //props access
    const nameFunction = ( props ) => {       
      <>
        <h1>{ props.title }</h1>
      </>
    }

  //props access with destructuring
    const TitleFunction = ( { title } ) => {
      <>
        <h1>{ title }</h1>
      </>
    }
  
    export const FirstApp = ( { title, subtitle } ) => {
      return (
        <>
            <h1>{ title }</h1> 
            <p>{ subtitle + 25000 }</p>
        </> 
      )
    }

      // default value of props 
        // export const FirstApp = ( { title = 'Hola mundo' } ) => {}

//lesson 41 - propTypes
    //this way the code generate an Error inside the component, it is not a good practice
  export const PropsAppErr = ( { title } ) => {
      if( !title ) {
        throw new Error( 'The title is missing' );
      }
      return (
        <h1>{ title }</h1>
      )
    }

    //Imported and used PropTypes, with Vite the PropTypes are not downlouded
      import PropTypes from 'prop-types';
      
      export const PropsApp = ( { title } ) => {
        return (
          <h1>{ title }</h1>
        )
      }
    
      PropsApp.protoTypes = {
        title: PropTypes.string.isRequired,
      }

//lesson 42 - DefaultProps 
  //DefaultProps in the top of the component
    // export const DefPropsApp = ({
    //   title,
    //   subtitle = 'Gromarant',
    // }) => {
    //     return (
    //      <>
    //         <h1>{ title }</h1>
    //         <h2>{ subtitle }</h2>
    //       </>
    //     )
    //   }


  //DefaultProps in the bottom of the component
    export const DefPropsApp = ({ title, subtitle, name }) => {
        return (
         <>
            <h1>{ title }</h1>
            <h2>{ subtitle }</h2>
            <h3>{ name }</h3>
          </>
        )
      }

    DefPropsApp.defaultProps = {
      name: 'Mariangelica Rodriguez',
      subtitle: 'There is not subtitle',
      title: 'There is not title',
    }

    DefPropsApp.protoTypes = {
      subtitle: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
    } 