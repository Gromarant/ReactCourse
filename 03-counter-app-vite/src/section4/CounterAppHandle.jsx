//lesson 45 handleSubtract and handleReset
  //Refactor version
  import { useState } from 'react';
  import PropTypes from 'prop-types';

  export const CounterAppHandle = ( { value } ) => {

    const [ counter, setCounter ] = useState( value );    //useState set the initial value
    const handleEvent = () => setCounter ( counter + 1 ); 
    const handleSubtract = () => setCounter( counter -1 );
    const handleReset = () => setCounter( value );              //setCounter( ( c ) => c + 1 );  c = current value of counter

    return (
      <>
        <h1>handleSubtract and handleReset</h1>
        <h2> { counter } </h2>
        <button onClick={ handleEvent }> +1 </button>
        <button onClick={ handleSubtract }> -1 </button>
        <button aria-label="btn-reset" onClick={ handleReset }> Reset </button>
      </>
    )
  }

  CounterAppHandle.propTypes = {
    value: PropTypes.number.isRequired,
  }