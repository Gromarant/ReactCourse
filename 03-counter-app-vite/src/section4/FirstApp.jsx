import Proptypes from 'prop-types';

export const FirstApp = ({ title, subTitle, name }) => {
  return (
    <>
        <h1 data-testid="test-title">{ title }</h1>
        <h3>{ subTitle }</h3>
        <h3>{ subTitle }</h3>
        <p>{ name }</p>
    </>
  )
}

FirstApp.Proptypes = {
    title: Proptypes.string.isRequired,
    subTitle: Proptypes.string,
}

FirstApp.defaultProps = {
    name: 'Mariangelica Rodriguez',
    subTitle: 'No hay subtítulo',
    // title: 'No hay título',
}