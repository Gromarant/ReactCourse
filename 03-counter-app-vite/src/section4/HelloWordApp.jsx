//  export function App() {
//     return <h1> Hola mundo =) !! </h1>;
// }

//Function with shortCut 'rafc'    
    export const HelloWordApp = () => {
      return (
        <h1>Hello Word App</h1>
      )
    }
    
