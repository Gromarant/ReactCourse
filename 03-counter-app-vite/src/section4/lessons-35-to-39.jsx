// lesson 35 - First component

import { Fragment } from "react";

    //rafc shorcut to create a component
        export const FirstComponent = ( ) => {        
            return (
                <h1>{ salaryOfert(50000) }</h1>
        )}
         
        //arrow function
            export const FirstArrowComponent = ( ) => <h1>{ salaryOfert(50000) }</h1>;
        
// lesson 36 - Homework - FirstApp
    export const FirstApp = ( ) => {
        return (
            <h1>Mariangelica</h1>
        )
    }

// lesson 37 - return elements inside the component - Fragments
export const NameApp = ( ) => {
    return (
        <> 
            <h1>Mariangelica</h1>
            <h2>Rodriguez</h2>
        </>
    )
}
        // <Fragment> </Fragment> ( it must be imported ) === <> </> ( it is not need be imported )
        //Fragment is use to return two elements

// lesson 38 - print variables inside the Html
    const salaryOfert = (salary) => {              
        return `1. The salary of ${ salary } Euros is for you`;
    };

    export const SalaryApp = ( ) => {
        return (
        <>         
            <h1>{ salaryOfert(50000) }</h1>
            <h2>2. Gromarant</h2>
        </>
        )
    }

    // Homework : add more examples and values
        const markName = '2. ';
        let addNumbers = '5. ' + 123;
        let addBoolean = true;
        const addArray = ['7. ', 1, 2, 3, 4, 5, 6, 7, 8, 9];
        const addObject = {
            number: '8. ',
            title: 'Mariangelica',
            message: 'Hola mundo',
        };

        export const Examples = ( ) => {
        return (
            <>
                <p> 3. Portafolio y página profesional de Mariangelica Rodríguez, Ux Front end Developer www.gromarant.com, donde muestra sus proyectos y proceso de desarrollo</p>
                <p>{ '4. ' + 2 * 8 }</p>      {/* directs operations  */}
                <p>{ addNumbers }</p>
                <p>{ addBoolean }</p>
                <p>{ addArray }</p>
                <p>{ JSON.stringify( addObject ) }</p>        {/* React do not accept objects to render, if you need print an object you must have use a 'JSON.stringify'*/}
            </>
        )
        }

// lesson 39 - add style.css
        // style.css file