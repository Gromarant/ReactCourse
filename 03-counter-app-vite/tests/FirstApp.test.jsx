import { getByText, render } from '@testing-library/react';
import { FirstApp } from '../src/section4/FirstApp';

describe('Test to <FirstApp />', () => {  

    test('1. Must be do match with the snapshot', () => {

        const title = 'Hola, Goku';
        const { container } =  render( <FirstApp title={ title } /> );
        
        expect( container ).toMatchSnapshot();
    });

    test('2. must be show the title at h1 tag', () => { 

        const title = 'Hola, Goku';
        const { container, getByText } =  render( <FirstApp title={ title } /> );
        expect( getByText(title) ).toBeTruthy();

        const h1 = container.querySelector( 'h1' );
        expect( h1.innerHTML ).toContain( title );

     })
});