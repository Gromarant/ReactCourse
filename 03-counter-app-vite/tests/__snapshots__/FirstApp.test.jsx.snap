// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to <FirstApp /> 1. Must be do match with the snapshot 1`] = `
<div>
  <h1>
    Hola, Goku
  </h1>
  <h3>
    No hay subt√≠tulo
  </h3>
  <p>
    Mariangelica Rodriguez
  </p>
</div>
`;
